### PR TITLE
[SPARK-46966][Python] Add UDTF API for 'analyze' method to indicate subset of input table columns to select

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR adds a UDTF API for the 'analyze' method to indicate subset of input table columns to select.

For example, this UDTF populates this 'select' list to indicate that Spark should only return two input columns from the input table: 'input' and 'partition_col':

```
from pyspark.sql.functions import AnalyzeResult, OrderingColumn, PartitioningColumn, SelectedColumn
from pyspark.sql.types import IntegerType, Row, StructType
class Udtf:
    def __init__(self):
        self._partition_col = None
        self._count = 0
        self._sum = 0
        self._last = None

    @staticmethod
    def analyze(row: Row):
        return AnalyzeResult(
            schema=StructType()
                .add("user_id", IntegerType())
                .add("count", IntegerType())
                .add("total", IntegerType())
                .add("last", IntegerType()),
            partitionBy=[
                PartitioningColumn("user_id")
            ],
            orderBy=[
                OrderingColumn("timestamp")
            ],
            select=[
                SelectedColumn("input"),
                SelectedColumn("partition_col")
            ])

    def eval(self, row: Row):
        self._partition_col = row["partition_col"]
        self._count += 1
        self._last = row["input"]
        self._sum += row["input"]

    def terminate(self):
        yield self._partition_col, self._count, self._sum, self._last
```

### Why are the changes needed?

This can reduce the amount of data sent between the JVM and Python interpreter, improving performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

This PR adds test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @ueshin

File: sql/core/src/test/scala/org/apache/spark/sql/IntegratedUDFTestUtils.scala [hunk 1]
```diff
@@ -627,22 +631,44 @@ object IntegratedUDFTestUtils extends SQLHelper {
...
       partitionBy = "partition_col",
-      orderBy = "partition_col ASC")
+      orderBy = "partition_col ASC",
+      select = "")
+
+  object UDTFInvalidSelectExprParseError
+    extends TestPythonUDTFPartitionByOrderByBase(
+      partitionBy = "partition_col",
+      orderBy = "input",
+      select = "\"unparsable\"")
```
Comment (ueshin) [inline sql/core/src/test/scala/org/apache/spark/sql/IntegratedUDFTestUtils.scala:671]: What happens if `select` contains only `partition_col`? That will try to get `row["input"]` in `eval`.
Could you add the test?

**Comment by ueshin** [comment]
I guess `"$partitionBy"` and `"$orderBy"` in the example of the descritpion should be some columns? Could you fix them?

File: sql/core/src/test/scala/org/apache/spark/sql/IntegratedUDFTestUtils.scala (hunk 1 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/IntegratedUDFTestUtils.scala]: Good idea, this is done. I added several more tests as well.

**Comment by dtenedor** [comment]
@ueshin thanks for your review! I switched the selected column representation to a new class `SelectedColumn` per offline suggestion, including an optional field for specifying an alias.

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.

File: python/pyspark/sql/worker/analyze_udtf.py [hunk 2]
```diff
@@ -31,7 +31,7 @@
     write_with_length,
     SpecialLengths,
 )
-from pyspark.sql.functions import PartitioningColumn
+from pyspark.sql.functions import OrderingColumn, PartitioningColumn, SelectedColumn
```
Comment (ueshin) [inline python/pyspark/sql/worker/analyze_udtf.py:34]:
```suggestion
from pyspark.sql.functions import PartitioningColumn, SelectedColumn
```

Comment (ueshin) [reply python/pyspark/sql/worker/analyze_udtf.py:34]: Or we may want a similar check for `OrderingColumn`? which can be done in a separate PR.

Comment (dtenedor) [reply python/pyspark/sql/worker/analyze_udtf.py]: Fixed this for now by removing `OrderingColumn`. I will add a similar check for that in a separate PR to help separate concerns.

**Comment by ueshin** [comment]
The failed tests seem not related to the last commit. The previous build passed except for linter that is fixed in the last commit.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**